### PR TITLE
Gh1-esb : Added rabbitmq import/export processors

### PR DESCRIPTION
**Pull request by haneloreianoseck**
metasfresh/metasfresh-edi#1

### CODE REVIEW

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+		int port = expProcessor.getPort();
+		String account = expProcessor.getAccount();
+		String password = expProcessor.getPasswordInfo();
+		String exchangeName = StringUtils.EMPTY;
+		String routingKey = StringUtils.EMPTY;
+		boolean isDurableQueue = true;
+
+		// Read all processor parameters and set them!
+		I_EXP_ProcessorParameter[] processorParameters = expProcessor.getEXP_ProcessorParameters();
+		if (processorParameters != null && processorParameters.length > 0)
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java]: avoid unneeded nesting; see if we can make sure that `expProcessor.getEXP_ProcessorParameters()` never returns `null` so we can remove this `if`

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+		{
+			for (I_EXP_ProcessorParameter processorParameter : processorParameters)
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java]: use a java-5 for-loop instead

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+				else if (processorParameter.getValue().equals(IS_DURABLE_QUEUE_PARAMETER))
+				{
+					isDurableQueue = Boolean.parseBoolean(processorParameter.getParameterValue());
+				}
+			}
+		}
+
+		if (exchangeName == null || exchangeName.length() == 0)
+		{
+			throw new Exception("Missing " + X_EXP_ProcessorParameter.Table_Name + " with key 'exchangeName'!");
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java]: pls introduce a little non-raw exception for this..like "ExportProcessorException"

edit: or maybe go with `ReplicationException`

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+		}
+
+		if (routingKey == null || routingKey.length() == 0)
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java]: use `Check.isEmpty(routingKey, true);` instead.
Or consider replacing the whole block with `Check.assumeNotEmpty(routingKey, ThatNewExportProcessorException.class, "Missing {} with key 'routingKey'!", X_EXP_ProcessorParameter.Table_Name)`

(`de.metas.util.Check`)

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+		{
+			throw new Exception("Missing " + X_EXP_ProcessorParameter.Table_Name + " with key 'routingKey'!");
+		}
+
+		// Construct Transformer Factory and Transformer
+		TransformerFactory tranFactory = TransformerFactory.newInstance();
+		String jVersion = System.getProperty("java.version");
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java]: no need to care for java 1.5 anymore

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+		Transformer aTransformer = tranFactory.newTransformer();
+		aTransformer.setOutputProperty(OutputKeys.INDENT, "yes");
+		Source src = new DOMSource(document);
+
+		// =================================== Write to String
+		Writer writer = new StringWriter();
+		Result dest2 = new StreamResult(writer);
+		aTransformer.transform(src, dest2);
+
+		sendJMSMessage(host, port, writer.toString(), exchangeName, routingKey, account, password, isDurableQueue);
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java]: update message name

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqImportProcessor.java [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+import org.slf4j.Logger;
+
+import java.text.MessageFormat;
+import java.util.List;
+import java.util.Properties;
+
+/**
+ * Aim of this class is to import records from AMQP Server.
+ */
+public class RabbitMqImportProcessor implements IImportProcessor
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqImportProcessor.java]: The same comments I had in `RabbitMqExportProcessor` :-)

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+import org.adempiere.server.rpl.api.IIMPProcessorBL;
+import org.adempiere.server.rpl.api.IImportHelper;
+import org.adempiere.server.rpl.exceptions.ReplicationException;
+import org.adempiere.util.lang.Mutable;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.springframework.amqp.core.*;
+import org.springframework.amqp.rabbit.connection.CachingConnectionFactory;
+import org.springframework.amqp.rabbit.core.RabbitAdmin;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java]:
```suggestion
```

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+import org.w3c.dom.Document;
+
+import javax.naming.Binding;
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java]:
```suggestion
```

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqImportProcessor.java [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+ */
+
+import de.metas.logging.LogManager;
+import de.metas.util.Services;
+import org.adempiere.server.rpl.IImportProcessor;
+import org.adempiere.server.rpl.IReplicationProcessor;
+import org.adempiere.server.rpl.api.IIMPProcessorBL;
+import org.adempiere.server.rpl.api.IIMPProcessorDAO;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.xmlbeans.impl.tool.Extension;
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqImportProcessor.java]:
```suggestion
```

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+/**
+ * Listen for AMQP Messages
+ */
+public class RabbitMqListener implements MessageListener
+{
+
+	/**
+	 * connection factory to AMQP server
+	 */
+	private CachingConnectionFactory connectionFactory;
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java]: ..same comment regarding unneeded comments/javadoc :-)

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+				"queueName=" + queueName + ", consumerTag=" + consumerTag, // reference
+				null); // exception
+	}
+
+	/**
+	 * Import xml document
+	 *
+	 * @param xml xml document (as string) to be imported
+	 */
+	private void importXMLDocument(final String xml)
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java]: ..unneeded javadoc; consider using `@NonNull`

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+		Services.get(ITrxManager.class).run(trxName, trxName -> {
+			final IImportHelper impHelper = Services.get(IIMPProcessorBL.class).createImportHelper(ctx);
+			try
+			{
+				final StringBuilder result = new StringBuilder();
+				responseDOM.setValue(impHelper.importXMLDocument(result, documentToBeImported, trxName));
+			}
+			catch (final Exception e)
+			{
+				throw e instanceof AdempiereException ? (AdempiereException)e : new AdempiereException(e);
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java]:
```suggestion
				throw AdempiereException.wrapIfNeeded(e);
```

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+			isStopping = false;
+
+			// 02486 make sure that we inform 'replicationProcessor' so there is a change of being restarted
+			replicationProcessor.setProcessRunning(false);
+		}
+	}
+
+	private void logException(final Throwable e, final String messageText)
+	{
+		log("Error: " + e.getLocalizedMessage(),
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java]: I think we can use ILoggable to the logging to both loggers and import processor log

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java (hunk 3 above)
Comment (teosarca) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java]: and always use not checked exceptions

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java (hunk 8 above)
Comment (teosarca) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java]: i would also avoid "*" imports.
I think IDEs are good enough to make it easy to precisely import what u need.

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java (hunk 2 above)
Comment (haneloreianoseck) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java]: did you mean another type of for-loop? This is a java-5 for-loop

Comment (metas-ts) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/exp/RabbitMqExportProcessor.java]: @haneloreianoseck right, sry

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java (hunk 14 above)
Comment (metas-ts) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java]: in order to use `ILoggable` afaiu you need to 
* create an `ILoggable` implementation that delegates to `Services.get(IIMPProcessorBL.class).createLog();`
* find a place where you can call `Loggables.temporarySetLoggable(ILoggable)` such that the BL that is running during import and export get's your ILogger

Comment (metas-ts) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java]: @haneloreianoseck pls check out this commit: https://github.com/metasfresh/metasfresh/commit/7e57d975e44cb7a45f89933912610b66b627fb61
wdyt?

Comment (metas-ts) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/server/rpl/imp/RabbitMqListener.java]: Note:
> find a place where you can call Loggables.temporarySetLoggable(ILoggable) such that the BL that is running during import and export get's your ILogger

on export we don't log to and sortof audit table..right? so there afais using ILoggable doesn't make sense

**Pull request merged**